### ISSUE TITLE
`extract_multiple` doesn't always extract as many variations as requested

### ISSUE THREAD

**Issue by ricardoV94** [egraphs-good/egglog#300]
Even if more exist as evidenced by testing via `egraph.check`. I can try to dig up a MWE if needed.

**Comment by yihozhang** [comment]
Ah right, this is a known behavior of egglog's variant extraction. To give some background, the feature is ad hoc and was initially developed to just sample a set of expressions from the database for debugging. AFAIR, the implementation tries to cover a diverse set of represented expressions (for debugging purposes) so it skips some expressions during the extraction, which causes the behavior described in this issue.

**Comment by ricardoV94** [comment]
Would it be useful to have something that works as I expected it to :D ?

**Comment by oflatt** [comment]
Definitely! It would be cool if `extract-variants` could recur after exhausting all of the enodes at the top level.

**Comment by Qi-Zhan** [comment]
If the egraph is acyclic, is it possible to write a function to extract all variants at the top level?
I think it will be a good feature, as sometimes we need to get all possible results to analyze further

**Comment by oflatt** [comment]
Yes, that would be possible
We agree that would be a good feature- it's something we should look into implementing in a couple months, once the new backend for egglog is stable
